runtime/lua/vim/lsp.lua

```diff
@@ -402,27 +401,36 @@ do
       if state.use_incremental_sync then
         -- This must be done immediately and cannot be delayed
         -- The contents would further change and startline/endline may no longer fit
-        table.insert(state.pending_changes, incremental_changes(client))
+        if not state.pending_changes[uri] then
+          state.pending_changes[uri] = {}
+        end
+        table.insert(state.pending_changes[uri], incremental_changes(client))
       end
       state.pending_change = function()
         state.pending_change = nil
         if client.is_stopped() or not vim.api.nvim_buf_is_valid(bufnr) then
           return
         end
-        local contentChanges
         if state.use_incremental_sync then
-          contentChanges = state.pending_changes
+          for change_uri, content_changes in pairs(state.pending_changes) do
+            client.notify("textDocument/didChange", {
+              textDocument = {
+                uri = change_uri;
+                version = util.buf_versions[vim.uri_to_bufnr(change_uri)];
+              };
+              contentChanges = content_changes,
+            })
+          end
           state.pending_changes = {}
         else
-          contentChanges = { full_changes(), }
+          client.notify("textDocument/didChange", {
```

Comment:
Will this not result in the reported issue if incremental sync is disable? Doesn't the `if` above which adds to `pending_changes` be rewritten as
```lua
      if state.use_incremental_sync then
        if not state.pending_changes[uri] then
          state.pending_changes[uri] = {}
        end
        table.insert(state.pending_changes[uri], incremental_changes(client))
      else
        state.pending_changes[uri] = true
      end
```
or something like that, and this one to
```lua
        -- drop the outer `if state.use_incremental_sync`
        for change_uri, content_changes in pairs(state.pending_changes) do
          if state.use_incremental_sync then
            content_changes = full_changes()
          end
          client.notify("textDocument/didChange", {
            textDocument = {
              uri = change_uri;
              version = util.buf_versions[vim.uri_to_bufnr(change_uri)];
            };
            contentChanges = content_changes,
          })
        end
        state.pending_changes = nil
```

Reply:
@mjlbach Could you please take a look?

Reply:
If you can provide an example that breaks (or even better, produce a test case) I can have a closer look.

Reply:
Nevermind, I checked the repro posted in the original instance with the exact same settings, but `allow_incremental_sync` disabled, and, apparently, it doesn't show up.

Reply:
I agree the way I chose to implement it is kinda messy and probably could be cleaned up to make more explicit